packages/go_router/lib/src/delegate.dart

```diff
@@ -150,22 +150,46 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
   }
 
   bool _onPopPage(Route<Object?> route, Object? result) {
-    if (!route.didPop(result)) {
-      return false;
+    if (route.willHandlePopInternally) {
+      final bool popped = route.didPop(result);
+      assert(!popped);
+      return popped;
     }
     final Page<Object?> page = route.settings as Page<Object?>;
     final RouteMatch? match = builder.getRouteMatchForPage(page);
     assert(match != null);
+    if (match!.route.onExit != null) {
+      // The `onExit` may perform updates to the navigator, such as popping up a
+      // dialog. Using a microtask to ensure the navigator finishes the current
+      // update before calling the `onExit`.
+      scheduleMicrotask(() async {
+        final FutureOr<bool> onExitResult =
+            match.route.onExit!(navigatorKey.currentContext!);
+        if ((onExitResult is bool && route.didPop(result)) ||
```

Comment:
Now I think of it, do you really need to call didPop if you call `_removeMatchFromList` ?

There is no way to wait for result for regular route match, you can only wait for ImperativeRouteMatch which handled in line 185

Reply:
I don't think so. So can I just `await onExitResult`?

Reply:
yeah I think so

Reply:
Done in [Await onExit](https://github.com/flutter/packages/pull/3709/commits/28dcce0d53dc79cd932e7917701ebe9ad498e5ed)